### PR TITLE
Session-based auth flow along with kill functionality

### PR DESCRIPTION
**Pull request by bjgajjar**

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
Couple of questions so far:

* Do we need both `access_token` and `PHPSESSID`?
* Can we rename `access_token` to `session`?
* Does the API re-set the cookie when using it? [1]
* Does the API auto-flush expired cookies from the user_sessions table?

File: src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php [hunk 1]
```diff
@@ -83,13 +83,11 @@ public function __invoke(Request $request, Response $response, callable $next)
...
-                        $cookie = new Cookies();
-                        $cookie->set('session',['value' => $authorizationTokenObject['token'],'expires' =>$expiry->format(\DateTime::COOKIE),'path'=>'/','httponly' => true]);
-                        $response =  $response->withAddedHeader('Set-Cookie',$cookie->toHeaders());
-                    }else{
-                        $response =  $response->withoutHeader('Set-Cookie');
-                    }
+                    $cookie = new Cookies();
+                    $expiryAt = $userSession ? $expiry->format(\DateTime::COOKIE) : DateTimeUtils::now()->toString();
+                    $cookie->set('session',['value' => $authorizationTokenObject['token'],'expires' => $expiryAt ,'path'=>'/','httponly' => true]);
+                    $cookie->set('PHPSESSID',['expires' => DateTimeUtils::now()->toString()]);
```
Comment (rijkvanzanten) [inline src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php]: Do we need this @bjgajjar? Why do we have this? What does it do?

Comment (rijkvanzanten) [reply src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php]: `PHPSESSID` that is

Comment (bjgajjar) [reply src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php]: - We need this to update the `expired_at` time of `session` cookie in every `API` call.
- As you have recommended we don't need `PHPSESSID` as a cookie. So trying to remove it from the response.

**Comment by rijkvanzanten** [comment]
Does this set the cookie `path` based on the project used?

We should double check if you can be logged into multiple projects on the same API at the same time

File: src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php [hunk 2]
```diff
@@ -96,7 +96,8 @@ public function __invoke(Request $request, Response $response, callable $next)
         if(isset($userSession)){
             $userSessionService->update($userSession['id'],['token_expired_at' => $expiry->format('Y-m-d H:i:s')]);
         }
-       
+        $response = $response->withHeader('Access-Control-Allow-Origin', $request->getHeader('Origin'))
+        ->withHeader('Access-Control-Allow-Credentials', 'true');
```
Comment (rijkvanzanten) [inline src/core/Directus/Application/Http/Middleware/ResponseCacheMiddleware.php]: This should probably be fetched from the config file:

https://github.com/directus/api/blob/develop/config/api_sample.php#L116

and be set to `true` by default

**Pull request merged**